app/app.go

```diff
@@ -0,0 +1,108 @@
+	// or the App method is not implemented for the provided Kind.
+	// Typically, the App's ManifestData should indicate this as well.
+	ErrNotImplemented = errors.New("not implemented")
+)
+
+// ConversionRequest is a request to convert a Kind from one version to another
+type ConversionRequest struct {
+	SourceGVK schema.GroupVersionKind
+	TargetGVK schema.GroupVersionKind
+	Raw       RawObject
+}
+
+// RawObject represents the raw bytes of the object and its encoding, optionally with a decoded version of the object,
+// which may be any valid resource.Object implementation.
+type RawObject struct {
+	Raw      []byte                `json:",inline"`
+	Object   resource.Object       `json:"-"`
+	Encoding resource.KindEncoding `json:"-"`
+}
+
+// SubresourceRequest is a request to a custom subresource
+type SubresourceRequest struct {
+	ResourceIdentifier resource.FullIdentifier
+	SubresourcePath    string
+	Method             string
+	Headers            http.Header
+	Body               []byte
+}
+
+// Config is the app configuration used in a Provider for instantiating a new App.
+// It contains kubernetes configuration for communicating with an API server, the App's ManifestData as fetched
+// by the runner, and additional arbitrary configuration details that may be app-specific.
+type Config struct {
+	// KubeConfig is a kubernetes rest.Config used to communicate with the API server where the App's Kinds are stored.
+	KubeConfig rest.Config
+	// ManifestData is the fetched ManifestData the runner is using for determining app kinds and capabilities.
+	ManifestData ManifestData
+	// ExtraConfig is additional configuration information, which may be app or runner specific
+	ExtraConfig map[string]any
+}
+
+// Provider represents a type which can provide an app manifest, and create a new App when given a configuration.
+// It should be used by runners to determine an app's capabilities and create an instance of the app to run.
+type Provider interface {
+	// Manifest returns a Manifest, which may contain ManifestData or may point to a location where ManifestData can be fetched from.
+	// The runner should use the ManifestData to determine app capabilities.
+	Manifest() Manifest
+	// NewApp creates a new App instance using the provided config, or returns an error if an App cannot be instantiated.
+	NewApp(Config) (App, error)
+}
+
+// Runnable represents a type which can be run until it errors or the provided channel is stopped (or receives a message)
+type Runnable interface {
+	// Run runs the process and blocks until one of the following conditions are met:
+	// * An unrecoverable error occurs, in which case it returns the error
+	// * The provided channel is closed, in which case processing should stop and the method should return
+	// * The provided channel is sent a message, in which case processing should stop and the method should return
+	// * The process completes and does not need to run again
+	Run(<-chan struct{}) error
```

Comment:
I'm fine with that, this was mainly modeled after the way kubernetes handles things so that the channel could be passed around internally, but we can always translate into a channel for the k8s stuff.